src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1802,73 +1856,90 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
+                    // source->target and target->source entities because database metadata hasn't yet been pulled to confirm which combination
+                    // of optimistically generated FK defintions matched the database FK relationship metadata.
                     //
-                    // So, before concluding that there is no FK constraint between the source, target entities, we need
-                    // to confirm absence of FK constraint from source->target and target->source tables.
-                    RelationShipPair inverseFKPair = new(fKDefinitionToTarget.Pair.ReferencedDbTable, fKDefinitionToTarget.Pair.ReferencingDbTable);
-
-                    // Add FK definition to the set of validated FKs only if no FK constraint is defined for the source and target entities
-                    // in the database, either from source -> target or target -> source.
-                    // When a foreign key constraint is not identified using inverseFKPair and fKDefinitionToTarget.Pair, it means that
-                    // the FK constraint is only defined in the runtime config file.
-                    if (PairToFkDefinition is not null && !PairToFkDefinition.ContainsKey(inverseFKPair))
+                    // At this point in the code, DAB has the database resolved FK metadata is available and can decide whether
+                    // 1. configResolvedFkDefinition has a matched database fk definition and isn't added to the list of
+                    //    validated FK definitions because it's already added.
+                    // 2. configResolvedFkDefinition doesn't have a matched database fk definition and is added to the list of
+                    //    validated FK definitions because it's not already added.
+                    bool doesFkExistInDatabase = VerifyForeignKeyExistsInDB(
+                        databaseTableA: configResolvedFkDefinition.Pair.ReferencingDbTable,
+                        databaseTableB: configResolvedFkDefinition.Pair.ReferencedDbTable);
+
+                    if (!doesFkExistInDatabase)
                     {
-                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
+                        validatedFKDefinitionsToTarget.Add(configResolvedFkDefinition);
+
+                        // The following operation generates FK metadata for use when processing requests on self-joined/referencing entities.
+                        EntityRelationshipKey key = new(entityName: configResolvedFkDefinition.SourceEntityName, configResolvedFkDefinition.RelationshipName);
+                        if (!RelationshipToFkDefinitions.TryAdd(key, configResolvedFkDefinition))
+                        {
+                            Console.WriteLine("for config-only defined relationship, no db fk, dab is failing to keep track of both 2 generated fk definitions.");
```

Comment:
nit: update L1941 to throw an exception.

Reply:
went with using "tryadd" and conditioned this section (adding entityrelationship keys) to self-join scenario only.